jscomp/syntax/src/res_core.ml

```diff
@@ -2767,34 +2769,39 @@ and parseJsxProp p =
         in
         Some (label, attrExpr))
   (* {...props} *)
-  | Lbrace -> (
+  | Lbrace ->
+    Scanner.popMode p.scanner Jsx;
     Parser.next p;
-    match p.Parser.token with
-    | DotDotDot -> (
-      Scanner.popMode p.scanner Jsx;
-      Parser.next p;
-      let loc = mkLoc p.Parser.startPos p.prevEndPos in
-      let propLocAttr =
-        (Location.mkloc "res.namedArgLoc" loc, Parsetree.PStr [])
-      in
-      let attrExpr =
-        let e = parsePrimaryExpr ~operand:(parseExpr p) p in
-        {e with pexp_attributes = propLocAttr :: e.pexp_attributes}
-      in
-      (* using label "spreadProps" to distinguish from others *)
-      let label = Asttypes.Labelled "_spreadProps" in
+    let props =
       match p.Parser.token with
-      | Rbrace ->
+      | DotDotDot -> (
         Parser.next p;
-        Some (label, attrExpr)
-      | _ -> None)
-    | _ -> None)
+        let loc = mkLoc p.Parser.startPos p.prevEndPos in
+        let propLocAttr =
+          (Location.mkloc "res.namedArgLoc" loc, Parsetree.PStr [])
+        in
+        let attrExpr =
+          let e = parsePrimaryExpr ~operand:(parseExpr p) p in
+          {e with pexp_attributes = propLocAttr :: e.pexp_attributes}
+        in
+        (* using label "spreadProps" to distinguish from others *)
+        let label = Asttypes.Labelled "_spreadProps" in
+        match p.Parser.token with
+        | Rbrace ->
+          Parser.next p;
+          Some (label, attrExpr)
+        | _ -> None)
+      | _ -> None
+    in
+    Scanner.setJsxMode p.scanner;
+    props
   | _ -> None
 
 and parseJsxProps p =
   parseRegion ~grammar:Grammar.JsxAttribute ~f:parseJsxProp p
 
 and parseJsxChildren p =
+  Scanner.popMode p.scanner Jsx;
```

Comment:
JSX children should not be parsed in JSX mode because they could be any kind of expression between curly braces.